pygfx/utils/transform.py

```diff
@@ -0,0 +1,555 @@
+
+    @property
+    def scale_x(self) -> float:
+        return self.scale[0]
+
+    @property
+    def scale_y(self) -> float:
+        return self.scale[1]
+
+    @property
+    def scale_z(self) -> float:
+        return self.scale[2]
+
+    @property
+    def right(self):
+        return self._directions[0]
+
+    @property
+    def up(self):
+        return self._directions[1]
+
+    @property
+    def forward(self):
+        return self._directions[2]
+
+    @position.setter
+    def position(self, value):
+        self.matrix = la.matrix_make_transform(value, self.rotation, self.scale)
+
+    @rotation.setter
+    def rotation(self, value):
+        self.matrix = la.matrix_make_transform(self.position, value, self.scale)
+
+    @scale.setter
+    def scale(self, value):
+        self.matrix = la.matrix_make_transform(self.position, self.rotation, value)
+
+    @gravity.setter
+    def gravity(self, value):
+        self._gravity = np.asarray(value)
+        self.flag_update()
+
+    @x.setter
+    def x(self, value):
+        _, y, z = self.position
+        self.position = (value, y, z)
+
+    @y.setter
+    def y(self, value):
+        x, _, z = self.position
+        self.position = (x, value, z)
+
+    @z.setter
+    def z(self, value):
+        x, y, _ = self.position
+        self.position = (x, y, value)
+
+    @scale_x.setter
+    def scale_x(self, value):
+        _, y, z = self.scale
+        self.scale = (value, y, z)
+
+    @scale_y.setter
+    def scale_y(self, value):
+        x, _, z = self.scale
+        self.scale = (x, value, z)
+
+    @scale_z.setter
+    def scale_z(self, value):
+        x, y, _ = self.scale
+        self.scale = (x, y, value)
+
+    @right.setter
+    def right(self, value):
+        value = np.asarray(value)
+
+        # make X look at the value by making Z look at the value and
+        # then rotate around Y by -pi/2
+        partA = la.matrix_make_look_at((0, 0, 0), value, up_reference=-self._gravity)
+        partA = la.matrix_to_quaternion(partA)
+        partB = la.quaternion_make_from_axis_angle((1, 0, 0), -np.pi / 2)
+        rotation = la.quaternion_multiply(partB, partA)
+
+        self.rotation = la.quaternion_multiply(rotation, self.rotation)
+
+    @up.setter
+    def up(self, value):
+        value = np.asarray(value)
+
+        # make Y look at the value by making Z look at the value and
+        # then rotate around X by -pi/2
+        partA = la.matrix_make_look_at((0, 0, 0), value, up_reference=-self._gravity)
+        partA = la.matrix_to_quaternion(partA)
+        partB = la.quaternion_make_from_axis_angle((0, 1, 0), -np.pi / 2)
+        rotation = la.quaternion_multiply(partB, partA)
+
+        self.rotation = la.quaternion_multiply(rotation, self.rotation)
+
+    @forward.setter
+    def forward(self, value):
+        rotation = la.matrix_make_look_at((0, 0, 0), value, up_reference=-self._gravity)
+        rotation = la.matrix_to_quaternion(rotation)
```

Comment:
Should we also check for the case where value is all zeros?

Also, would it make sense for `WorldObject.look_at()` to simply set this prop?

Reply:
> Should we also check for the case where value is all zeros?

Good catch. I keep getting confused by our choice of naming the parent frame `local`. Parameters are set in the parent's frame (which is `target` not `source`). Thus the origin (`eye`) mustn't always be `(0, 0, 0)` and should be `self.position` instead, which is the origin of the source frame (local) expressed in the target frame (parent).

In this case, a check makes a lot of sense, because the not allowed value depends on the position of the frame.

> Also, would it make sense for WorldObject.look_at() to simply set this prop?

Yes. We could even drop the entire function since `WorldObject.look_at(value)` is the same as `WorldObject.world.forward = value` and Python has a "one way to do a thing" philosophy. At the same time, we have the "if ThreeJS does it we probably want to do it, too" philosophy, so we might want to still keep `look_at` around?

Reply:
I would agree to keep it around, but make it set the property like Almar suggested.

Reply:
I can't get over how amazingly elegant this API has turned out, @FirefoxMetzger ! <3

> `ob.world.forward = value - ob.world.position`

That reads like poetry to me.

Reply:
> I think ob.look_at(value) would translate to ob.world.forward = value - ob.world.position, or am I missing something?

We are currently doing this internally. `obj.world.forward` sets the world rotation in such a way that the object's forward axis **points towards `value`**.

Maybe I misunderstood what it should do and it should instead be "`obj.world.forward` sets the world rotation in such a way that the object's forward axis **expressed in world space equals `value` (up to scale)**"?

Reply:
On second thought, my second definition makes more sense. I will update the PR (and `look_at`)